contentcuration/search/viewsets/contentnode.py

```diff
@@ -98,7 +98,7 @@ def filter_keywords(self, queryset, name, value):
         filter_query = Q(title__icontains=value) | Q(description__icontains=value)
         tags_node_ids = ContentNode.tags.through.objects.filter(
             contenttag__tag_name__icontains=value
-        ).values_list("contentnode_id", flat=True)
+        ).values_list("contentnode_id", flat=True)[:100]
```

Comment:
I'm not sure if this will interfere with the pagination results (e.g. # of pages). Would it make sense to update on the js side with something like `Math.min(100, params.page_size)` instead?

Reply:
The issue is that letting this query return an arbitrary number of results can result in incredibly poor performance on the server side, and when there are a large number of results, they are also not very diagnostic (i.e. you'd get this many results for searching by `Math`).

Reply:
as an example, if you search for 'math' you'll have 58000 results only from the search in tags, if you add serching in description and title it will be several hundred of thousands. Just removing duplicates and grouping them will take  hours if not days, not mentioning annotating some of the fields that are shown in the results...

Reply:
Why is it hardcoded to the first hundred results though? Shouldn't it be `[offset:limit]`, where `offset` is something like ` (page_num - 1) * page_size` and `limit` is `offset + page_size`? Otherwise queries with more than 100 total results can never return results past the hundredth one, can it?

Reply:
Because these are not the 100 first results, if they were, things could easily be solved with pagination :(

Search of the keyword is done in `tags` , `descriptions` and `titles`. This is the tags part, afterwards results from description and titles are added, then duplicated are removed, then all these results are grouped, and that's the final amount . This last query will take days if it contains hundred of thousands of results to be deduplicate and grouped. This PR just reduces this amount to a maximum of 200 that takes less than 30 seconds.

This is just an arbritary number that showed to work . I can't think of any way of limiting tags in advanced, taking into account the whole process.

Reply:
@jredrejo Right, I understand these aren't the first 100 results that will be returned to the user, but these are restricting tag queries to never returning more than 100 results, aren't they? So if tag result 103 is what the user really wanted, they will never see it no matter how many pages of results are returned, if I understand correctly.

I think the problem here is that we're trying to group disparate searches into a result set, which means we're essentially using a single filter as a series of filters, and that in turn causes us to have to prematurely execute queries in order to combine the filter results. In a sense, it feels like we're trying to use filters to implement a form of full text search.

Do we still have this problem if we separate out tag search from this? i.e. search by keyword becomes tag search, and we have a separate filter like description search (which could include title). If that works, what we could do then is actually run two queries (maybe 25 results each) then aggregate them in the frontend to return 50 results. (Ugly, I know, but we're in string freeze IIRC.) The grouping may be different from the current approach, but it should be consistent and all results can eventually be viewed should the user want to do so. 

I think after release we want to either separate them out as two separate search fields, or ideally just work on implementing real full text search. (I do have some ideas on that. :)

Reply:
> @jredrejo Right, I understand these aren't the first 100 results that will be returned to the user, but these are restricting tag queries to never returning more than 100 results, aren't they? So if tag result 103 is what the user really wanted, they will never see it no matter how many pages of results are returned, if I understand correctly.
> 

Yes, you're right 

> I think the problem here is that we're trying to group disparate searches into a result set, which means we're essentially using a single filter as a series of filters, and that in turn causes us to have to prematurely execute queries in order to combine the filter results. In a sense, it feels like we're trying to use filters to implement a form of full text search.
> 
> Do we still have this problem if we separate out tag search from this? i.e. search by keyword is only tag search, and we have a separate filter like description search (which could include title). If that works, what we could do then is actually run two queries (maybe 25 results each) then aggregate them in the frontend to return 50 results. (Ugly, I know, but we're in string freeze IIRC.) The grouping may be different from the current approach, but it should be consistent and all results can eventually be viewed should the user want to do so.

yes, we have still the problem, because the problem is not the actual search but the representation of the results. We've been able to improve the search performance, so getting a list of all the nodes ids that have the result is quite fast, in the worst cases is less than 10 seconds.

The problem comes when we represent the results, because of two reasons:
1. We are grouping the nodes in topics, so pagination of the nodes does not work, and grouping hundred of thousands of nodes takes a long time
2. We are adding a lot of information with the results, information that's being fetched from different tables with aggregations, and that takes a long time too.

So, mainly 1 but also 2, because some of the grouping conditions are aggregations is what is causing the problem, with two sides: on one side the speed of the final queryset is bad and on the other, there's not a way to paginate if you don't group first. If we could paginate without grouping, the problem is solved because we can modify the paginator to do the annotations only on the node ids needed per page.


> 
> I think after release we want to either separate them out as two separate search fields, or ideally just work on implementing real full text search. (I do have some ideas on that. :)

I would really like to know those ideas, but, with our current method of annotating + grouping + annotating I only can think of having a denormalized table (kind of Google Bigtable) to be able to provide the user with thousands of results after a search. But, from an user point of view: does it make sense? have you ever gone to the 10th page of Google results to find the good site? I am quite sure that telling the user he is at the page 1 of 8765 in his results will not be a good solution
I think that giving him the first 200 results, and maybe telling him something like "_your query resulted in too many results, please refine your criteria and try again_". would be better from an UX point of view.

Reply:
> 1. We are grouping the nodes in topics, so pagination of the nodes does not work, and grouping hundred of thousands of nodes takes a long time

Is this absolutely necessary for release? If we just stopped doing this, how much faster does that make the search? We already have the "Go to location" option in the search results if they want to see the topic the content is in, and this gives them the ability to see the content item in full context.

> 2. We are adding a lot of information with the results, information that's being fetched from different tables with aggregations, and that takes a long time too.

Similarly, is all this data necessary just to display the results? When I do a search, I see the thumbnail, title, description, and "go to location". (There's an info button but it seems like it needs to fetch data?) Or can some of this be deferred to when the user actually selects or investigates the result?

I'm okay with doing the limiting as a short-term measure to get something out the door, but one thing to note is that after aggregation, grouping, and sorting, tag query result 103 (or even 756) might end up actually being search result 5. So I think this is a problematic band-aid for performance issues, and I am concerned about this being something we rely on as a semi-permanent solution.

> But, from an user point of view: does it make sense? have you ever gone to the 10th page of Google results to find the good site? I am quite sure that telling the user he is at the page 1 of 8765 in his results will not be a good solution

I have, though quite rarely. I have done things like this in the Studio admin though, like searching for a word in a channel name, or searching all our sushi chef channels, then clicking through pages quickly to find the channel I'm looking for. In that case, the results could be on the last page, and that could be page 10 or even page 20. I try to avoid saying "the user would never want this or want to do this", because over time I think I've been proven wrong more often than not on that. ;)

Reply:
> > 1. We are grouping the nodes in topics, so pagination of the nodes does not work, and grouping hundred of thousands of nodes takes a long time
> 
> Is this absolutely necessary for release? If we just stopped doing this, how much faster does that make the search? We already have the "Go to location" option in the search results if they want to see the topic the content is in, and this gives them the ability to see the content item in full context.
> 
> > 1. We are adding a lot of information with the results, information that's being fetched from different tables with aggregations, and that takes a long time too.
> 
> Similarly, is all this data necessary just to display the results? When I do a search, I see the thumbnail, title, description, and "go to location". (There's an info button but it seems like it needs to fetch data?) Or can some of this be deferred to when the user actually selects or investigates the result?
> 
> I'm okay with doing the limiting as a short-term measure to get something out the door, but one thing to note is that after aggregation, grouping, and sorting, tag query result 103 (or even 756) might end up actually being search result 5. So I think this is a problematic band-aid for performance issues, and I am concerned about this being something we rely on as a semi-permanent solution.
> 

If we stop doing this, the results will be in a few seconds, even with hundred of thousands of results.
If you compare the way the results are shown with the current studio results, these are better, with more results and better organized and presented, so, when the number of results is reasonable, this is better. So I'd keep it , with few (<200) results. For more, it's causing a big problem, and I would not use it as it is, because it's causing server errors.

> > But, from an user point of view: does it make sense? have you ever gone to the 10th page of Google results to find the good site? I am quite sure that telling the user he is at the page 1 of 8765 in his results will not be a good solution
> 
> I have, though quite rarely. I have done things like this in the Studio admin though, like searching for a word in a channel name, or searching all our sushi chef channels, then clicking through pages quickly to find the channel I'm looking for. In that case, the results could be on the last page, and that could be page 10 or even page 20. I try to avoid saying "the user would never want this or want to do this", because over time I think I've been proven wrong more often than not on that. ;)

Agree, but it's rare, I'd say that whether we're facing a tradeoff, rare situations should be discarded as the first option. And, really, I doubt you'd be able to check 158000 results and stay awake :smile:

Reply:
> If we stop doing this, the results will be in a few seconds, even with hundred of thousands of results.
If you compare the way the results are shown with the current studio results, these are better, with more results and better organized and presented, so, when the number of results is reasonable, this is better. 

Is this something we objectively know to be true, or just our gut feeling on it? On the one hand, we're just randomly discarding most of the tag results, and on the other hand we're saying we get better search results when we do so. I can understand how less results feels less overwhelming, but it's hard to see how not including potentially thousands of results when we actually sort the results (that may have turned out to be the best picks in the final tally) can actually lead to better search results.

> Agree, but it's rare, I'd say that whether we're facing a tradeoff, rare situations should be discarded as the first option. And, really, I doubt you'd be able to check 158000 results and stay awake 😄

If you get 158000 results, you'll just do a more specific search, especially if it's fast! Also, keep in mind that Google has spent years perfecting predictive search algorithms so that you're highly likely to get the result you want on page one. Studio search doesn't have anywhere near this level of sophistication, so the results are not likely to be as good, especially given that we lack full-text search.

Anyway, it sounds like we might not come to an agreement on this. My concerns are on record, and it is true that getting any search results is an improvement over getting none, so this is definitely better than the current develop status quo.

Reply:
> > If we stop doing this, the results will be in a few seconds, even with hundred of thousands of results.
> > If you compare the way the results are shown with the current studio results, these are better, with more results and better organized and presented, so, when the number of results is reasonable, this is better.
> 
> Is this something we objectively know to be true, or just our gut feeling on it? On the one hand, we're just randomly discarding most of the tag results, and on the other hand we're saying we get better search results when we do so. I can understand how less results feels less overwhelming, but it's hard to see how not including potentially thousands of results when we actually sort the results (that may have turned out to be the best picks in the final tally) can actually lead to better search results.
> 


I did several searches in both servers (studio and develop.studio) , for the latest  using my local against the remote database to avoid the server protections timeouts. In all of them, develop provided same or more results (that were correct, I checked it) and were presented in a comfortable way, while current studio did even provide duplicates.


> > Agree, but it's rare, I'd say that whether we're facing a tradeoff, rare situations should be discarded as the first option. And, really, I doubt you'd be able to check 158000 results and stay awake smile
> 
> If you get 158000 results, you'll just do a more specific search, especially if it's fast! Also, keep in mind that Google has spent years perfecting predictive search algorithms so that you're highly likely to get the result you want on page one. Studio search doesn't have anywhere near this level of sophistication, so the results are not likely to be as good, especially given that we lack full-text search.
> 
> Anyway, it sounds like we might not come to an agreement on this. My concerns are on record, and it is true that getting any search results is an improvement over getting none, so this is definitely better than the current develop status quo.

Ah, don't misunderstand me: I do think this is not a right search and the results are not good when there are many occurences (they are correct if there are less than 200 nodes in the results), I don't like it but I think it's better than the current situation where the user sees an infinite spinner because the api request has been rejected by cloudfare. If we find a way to retrieve so many data in a reasonable time, I'll go for it, no doubt.

I'm thinking that for the case we have so many results we could skip grouping in topics, then we could play with the pagination and the queries would be fast because the annotations would be only over the nodes included in the page. That would mean that we have 2 ways of representing the search results, if that's considered a better option, I'd be happy to implement it while other solutions are found.

Reply:
> I'm thinking that for the case we have so many results we could skip grouping in topics, then we could play with the pagination and the queries would be fast because the annotations would be only over the nodes included in the page.  That would mean that we have 2 ways of representing the search results, if that's considered a better option, I'd be happy to implement it while other solutions are found. 

I personally would be much more comfortable with this approach, then we are not discarding anything, and if they get a bunch of results, they will probably just try a more specific search.